server/public/model/autotranslation.go

```diff
@@ -0,0 +1,116 @@
+// See LICENSE.txt for license information.
+
+package model
+
+import (
+	"encoding/json"
+	"io"
+	"maps"
+)
+
+// TranslationType indicates the type of translated content
+type TranslationType string
+
+const (
+	TranslationTypeString TranslationType = "string"
+	TranslationTypeObject TranslationType = "object"
+)
+
+// TranslationState represents the state of a translation
+type TranslationState string
+
+const (
+	TranslationStateReady       TranslationState = "ready"
+	TranslationStateSkipped     TranslationState = "skipped"
+	TranslationStateUnavailable TranslationState = "unavailable"
+)
+
+// Translation represents a single translation result
+type Translation struct {
+	ObjectID   string           `json:"object_id"`
+	ObjectType string           `json:"object_type"`
+	Lang       string           `json:"lang"`
+	Provider   string           `json:"provider"`
+	Type       TranslationType  `json:"type"`
+	Text       string           `json:"text"`
+	ObjectJSON json.RawMessage  `json:"object_json,omitempty"`
+	Confidence *float64         `json:"confidence,omitempty"`
+	State      TranslationState `json:"state"`
+	Meta       map[string]any   `json:"meta,omitempty"`
+	NormHash   *string          `json:"norm_hash,omitempty"`
+}
+
+func (t *Translation) ToJSON() string {
+	b, _ := json.Marshal(t)
+	return string(b)
+}
+
+func (t *Translation) FromJSON(data io.Reader) (*Translation, error) {
+	err := json.NewDecoder(data).Decode(&t)
+	if err != nil {
+		return nil, err
+	}
+	return t, nil
+}
+
+func (t *Translation) Clone() *Translation {
+	if t == nil {
+		return nil
+	}
+	var confidence *float64
+	if t.Confidence != nil {
+		val := *t.Confidence
+		confidence = &val
+	}
+	var normHash *string
+	if t.NormHash != nil {
+		val := *t.NormHash
+		normHash = &val
+	}
+	var meta map[string]any
+	if t.Meta != nil {
+		meta = make(map[string]any, len(t.Meta))
+		maps.Copy(meta, t.Meta)
+	}
+	var objectJSON json.RawMessage
+	if t.ObjectJSON != nil {
+		objectJSON = make([]byte, len(t.ObjectJSON))
+		copy(objectJSON, t.ObjectJSON)
+	}
+	return &Translation{
+		ObjectID:   t.ObjectID,
+		ObjectType: t.ObjectType,
+		Lang:       t.Lang,
+		Provider:   t.Provider,
+		Type:       t.Type,
+		Text:       t.Text,
+		ObjectJSON: objectJSON,
+		Confidence: confidence,
+		State:      t.State,
+		Meta:       meta,
+		NormHash:   normHash,
+	}
+}
+
+func (t *Translation) IsValid() bool {
+	if t == nil {
+		return false
+	}
+	if t.Provider == "" || t.ObjectID == "" || !IsValidId(t.ObjectID) || t.ObjectType == "" || t.Lang == "" || t.Type == "" || t.State != TranslationStateReady {
+		return false
+	}
+	if t.Type != TranslationTypeString && t.Type != TranslationTypeObject {
+		return false
+	}
+	if t.Type == TranslationTypeString && t.Text == "" {
+		return false
+	}
+	if t.Type == TranslationTypeObject && len(t.ObjectJSON) == 0 {
+		return false
+	}
+	if t.NormHash != nil && *t.NormHash == "" {
+		return false
+	}
```

Comment:
If `NormHash` is not empty, should we check that it is indeed the hash of the text?